zaza/openstack/charm_tests/ceph/rbd_mirror/tests.py

```diff
@@ -124,6 +136,119 @@ def wait_for_mirror_state(self, state, application_name=None,
                 # all images with state has expected state
                 return True
 
+    def get_cinder_rbd_mirroring_mode(self,
+                                      cinder_ceph_app_name='cinder-ceph'):
+        """Get the RBD mirroring mode for the Cinder Ceph pool.
+
+        :returns: A string representing the RBD mirroring mode. It can be
+                  either 'pool' or 'image'.
+        """
+        DEFAULT_RBD_MIRRORING_MODE = 'pool'
+
+        rbd_mirroring_mode_config = zaza.model.get_application_config(
+            cinder_ceph_app_name).get('rbd-mirroring-mode')
+        if rbd_mirroring_mode_config:
+            rbd_mirroring_mode = rbd_mirroring_mode_config.get(
+                'value', DEFAULT_RBD_MIRRORING_MODE).lower()
+        else:
+            rbd_mirroring_mode = DEFAULT_RBD_MIRRORING_MODE
+
+        return rbd_mirroring_mode
+
+    def create_cinder_volume(self, session, from_image=False):
+        """Create Cinder Volume from image.
+
+        :rtype: :class:`Volume`.
+        """
+        def get_glance_image(session):
+            glance = openstack.get_glance_session_client(session)
+            images = openstack.get_images_by_name(glance, CIRROS_IMAGE_NAME)
+            if images:
+                return images[0]
+            logging.info("Failed to find {} image, falling back to {}".format(
+                CIRROS_IMAGE_NAME,
+                LTS_IMAGE_NAME))
+            return openstack.get_images_by_name(glance, LTS_IMAGE_NAME)[0]
+
+        def create_volume_type(cinder):
```

Comment:
I think it should be `create_repl_volume_type_if_doesnt_exist` or so.

Also I'd rather have such generic reusable functions as private functions one level above (not as functions in functions) so that we could easily use them from somewhere else some day without having again to move the code (reviews where code gets moved are really time consuming for reviewers and require a lot of focus to try to spot tiny differences between a big red wall of code and a big green wall of code :D )

Reply:
Sounds good. I'll make the change.